packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,332 @@
+    } else {
+      // Otherwise, we don't know what device type we're on/
+      return DeviceType.UNKNOWN;
+    }
+  }
+
+  @ExpoMethod
+  public void getSerialNumberAsync(final Promise promise) {
+
+    int permissionCheck = ContextCompat.checkSelfPermission(mContext, Manifest.permission.READ_PHONE_STATE);
+    if (permissionCheck == PackageManager.PERMISSION_GRANTED) {
+      try {
+        if (android.os.Build.VERSION.SDK_INT < 26) {
+          promise.resolve(Build.SERIAL);
+        } else {
+          promise.resolve(Build.getSerial());
+        }
+      } catch (SecurityException e) {
+        Log.e(TAG, e.getMessage());
+        promise.reject(e);
+      }
+    } else {
+      Permissions mPermissions = mModuleRegistry.getModule(Permissions.class);
+      String permissionsTable[] = {Manifest.permission.READ_PHONE_STATE};
+      mPermissions.askForPermissions(permissionsTable, new Permissions.PermissionsRequestListener() {
+        @Override
+        public void onPermissionsResult(int[] results) {
+          if (results[0] == PackageManager.PERMISSION_GRANTED) {
+            try {
+              if (android.os.Build.VERSION.SDK_INT < 26) {
+                promise.resolve(Build.SERIAL);
+              } else {
+                promise.resolve(Build.getSerial());
+              }
+            } catch (SecurityException e) {
+              Log.e(TAG, e.getMessage());
+              promise.reject(e);
+            }
+          } else {
+            promise.reject(new SecurityException("User rejected permissions"));
+          }
+        }
+      });
+    }
+  }
+
+  @ExpoMethod
+  public void getFreeDiskStorageAsync(Promise promise) {
+    try {
+      StatFs external = new StatFs(Environment.getExternalStorageDirectory().getAbsolutePath());
+      long availableBlocks;
+      long blockSize;
+
+      availableBlocks = external.getAvailableBlocksLong();
+      blockSize = external.getBlockSizeLong();
+
+      BigInteger storage = BigInteger.valueOf(availableBlocks).multiply(BigInteger.valueOf(blockSize));
+      promise.resolve(storage.doubleValue());
+    } catch (NullPointerException e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getIpAddressAsync(Promise promise) {
+    try {
+      Integer ipAddress = getWifiInfo().getIpAddress();
+      // Convert little-endian to big-endianif needed
+      if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+        ipAddress = Integer.reverseBytes(ipAddress);
+      }
+      byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();
+      String ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getMACAddressAsync(String interfaceName, Promise promise) {
+    String permission = "android.permission.INTERNET";
```

Comment:
this should be a constant defined by the Android OS already

Reply:
The constant value is this string `"android.permission.INTERNET"` ( reference from https://developer.android.com/reference/android/Manifest.permission)